File: Zeus.Server.Hosting/Remote/Spake2Plus.cs
using System.Security.Cryptography;
using System.Text;
using Org.BouncyCastle.Asn1.Sec;
using Org.BouncyCastle.Asn1.X9;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Utilities.Encoders;
// Alias the BouncyCastle EC types — System.Security.Cryptography also defines
// ECPoint/ECCurve, and we need that namespace for SHA256/HKDF/HMAC.
using ECCurve = Org.BouncyCastle.Math.EC.ECCurve;
using ECPoint = Org.BouncyCastle.Math.EC.ECPoint;

namespace Zeus.Server.Hosting.Remote;

/// <summary>Which half of the SPAKE2+ exchange this instance plays.</summary>
public enum Spake2Role
{
    /// <summary>The client that knows the password (holds w0 + w1).</summary>
    Prover,
    /// <summary>The radio that stored the verifier (holds w0 + L). This is Zeus.Server.</summary>
    Verifier,
}

/// <summary>Thrown on any SPAKE2+ protocol/validation failure. Fails closed — leaks nothing.</summary>
public sealed class Spake2Exception(string message) : Exception(message);

/// <summary>
/// Outcome of a completed SPAKE2+ exchange. <see cref="SharedKey"/> is the
/// authenticated session key (K_shared / Ke); <see cref="LocalConfirm"/> is the
/// MAC this side sends, and <see cref="ExpectedPeerConfirm"/> is the MAC it must
/// receive (constant-time compared) before trusting the key.
/// </summary>
public sealed class Spake2PlusOutcome
{
    public required byte[] SharedKey { get; init; }
    public required byte[] LocalConfirm { get; init; }
    public required byte[] ExpectedPeerConfirm { get; init; }

    // Intermediates exposed only to tests (InternalsVisibleTo) for RFC 9383
    // Appendix C vector verification.
    internal byte[] Transcript { get; init; } = [];
    internal byte[] KMain { get; init; } = [];
    internal byte[] KConfirmP { get; init; } = [];
    internal byte[] KConfirmV { get; init; } = [];
    internal byte[] ConfirmP { get; init; } = [];
    internal byte[] ConfirmV { get; init; } = [];
    internal byte[] ShareP { get; init; } = [];
    internal byte[] ShareV { get; init; } = [];
    internal byte[] ZEnc { get; init; } = [];
    internal byte[] VEnc { get; init; } = [];
}

/// <summary>
/// SPAKE2+ (RFC 9383), ciphersuite P256-SHA256-HKDF-SHA256-HMAC-SHA256. The
/// augmented PAKE behind the remote-access session password (ADR-0008): the
/// server stores only the verifier (w0, L = w1·P); the client proves knowledge
/// of the password without ever sending it, and a malicious broker cannot MITM.
///
/// EC point math uses BouncyCastle (already in the dependency graph via
/// SIPSorcery); the KDF/MAC/hash use the BCL. Correctness is pinned to the RFC
/// 9383 Appendix C test vectors in <c>Spake2PlusTests</c>.
/// </summary>
public sealed class Spake2Plus
{
    private static readonly X9ECParameters Params = SecNamedCurves.GetByName("secp256r1");
    private static readonly ECCurve Curve = Params.Curve;
    private static readonly ECPoint G = Params.G;
    private static readonly BigInteger Order = Params.N;

    // RFC 9382/9383 constant points (compressed SEC1); discrete log unknown.
    private static readonly ECPoint M = Curve.DecodePoint(
        Hex.Decode("02886e2f97ace46e55ba9dd7242579f2993b64e16ef3dcab95afd497333d8fa12f"));
    private static readonly ECPoint N = Curve.DecodePoint(
        Hex.Decode("03d8bbd6c639c62937b04d997f38c3770719c629d7014d49a24b4f98baa1292b49"));

    private readonly Spake2Role _role;
    private readonly byte[] _context;
    private readonly byte[] _idProver;
    private readonly byte[] _idVerifier;

    private BigInteger? _w0;
    private BigInteger? _w1;   // prover only
    private ECPoint? _l;       // verifier only
    private BigInteger? _scalar; // x (prover) or y (verifier)
    private byte[]? _share;

    public Spake2Plus(Spake2Role role, byte[] context, byte[] idProver, byte[] idVerifier)
    {
        _role = role;
        _context = context;
        _idProver = idProver;
        _idVerifier = idVerifier;
    }

    /// <summary>Decode an uncompressed-SEC1 registration record L into a point.</summary>
    public static ECPoint DecodeL(byte[] encoded) => Validated(Curve.DecodePoint(encoded));

    /// <summary>
    /// Prover (client) first message: shareP = x·P + w0·M. <paramref name="x"/>
    /// is for test vectors only; production passes null for a random scalar.
    /// </summary>
    public byte[] StartProver(BigInteger w0, BigInteger w1, BigInteger? x = null)
    {
        if (_role != Spake2Role.Prover) throw new Spake2Exception("role mismatch");
        _w0 = w0;
        _w1 = w1;
        _scalar = x ?? RandomScalar();
        _share = G.Multiply(_scalar).Add(M.Multiply(w0)).Normalize().GetEncoded(false);
        return _share;
    }

    /// <summary>
    /// Verifier (radio) first message: shareV = y·P + w0·N. <paramref name="y"/>
    /// is for test vectors only; production passes null for a random scalar.
    /// </summary>
    public byte[] StartVerifier(BigInteger w0, ECPoint l, BigInteger? y = null)
    {
        if (_role != Spake2Role.Verifier) throw new Spake2Exception("role mismatch");
        _w0 = w0;
        _l = l;
        _scalar = y ?? RandomScalar();
        _share = G.Multiply(_scalar).Add(N.Multiply(w0)).Normalize().GetEncoded(false);
        return _share;
    }

    /// <summary>
    /// Process the peer's share, derive Z/V, build the transcript, and produce the
    /// session key + confirmation MACs. Throws (fails closed) on any invalid input.
    /// </summary>
    public Spake2PlusOutcome Process(byte[] peerShare)
    {
        if (_w0 is null || _scalar is null || _share is null)
            throw new Spake2Exception("Start* not called");

        var peer = Validated(Curve.DecodePoint(peerShare));

        ECPoint z, v;
        byte[] shareP, shareV;
        if (_role == Spake2Role.Prover)
        {
            // peer = shareV (Y). Z = x·(Y − w0·N); V = w1·(Y − w0·N).
            var t = Validated(peer.Subtract(N.Multiply(_w0)));
            z = t.Multiply(_scalar).Normalize();
            v = t.Multiply(_w1!).Normalize();
            shareP = _share;
            shareV = peer.GetEncoded(false);
        }
        else
        {
            // peer = shareP (X). Z = y·(X − w0·M); V = y·L.
            var t = Validated(peer.Subtract(M.Multiply(_w0)));
            z = t.Multiply(_scalar).Normalize();
            v = _l!.Multiply(_scalar).Normalize();
            shareP = peer.GetEncoded(false);
            shareV = _share;
        }
        if (z.IsInfinity || v.IsInfinity)
            throw new Spake2Exception("degenerate shared point");

        var zEnc = z.GetEncoded(false);
        var vEnc = v.GetEncoded(false);

        var tt = BuildTranscript(shareP, shareV, zEnc, vEnc, _w0!);
        var kMain = SHA256.HashData(tt);

        var confirmKeys = HKDF.DeriveKey(HashAlgorithmName.SHA256, kMain, 64,
            salt: null, info: Encoding.ASCII.GetBytes("ConfirmationKeys"));
        var kConfirmP = confirmKeys[..32];
        var kConfirmV = confirmKeys[32..];
        var kShared = HKDF.DeriveKey(HashAlgorithmName.SHA256, kMain, 32,
            salt: null, info: Encoding.ASCII.GetBytes("SharedKey"));

        // RFC 9383 §3.4: confirmP = MAC(K_confirmP, shareV); confirmV = MAC(K_confirmV, shareP).
        var confirmP = HMACSHA256.HashData(kConfirmP, shareV);
        var confirmV = HMACSHA256.HashData(kConfirmV, shareP);

        var (local, expected) = _role == Spake2Role.Prover
            ? (confirmP, confirmV)
            : (confirmV, confirmP);

        return new Spake2PlusOutcome
        {
            SharedKey = kShared,
            LocalConfirm = local,
            ExpectedPeerConfirm = expected,
            Transcript = tt,
            KMain = kMain,
            KConfirmP = kConfirmP,
            KConfirmV = kConfirmV,
            ConfirmP = confirmP,
            ConfirmV = confirmV,
            ShareP = shareP,
            ShareV = shareV,
            ZEnc = zEnc,
            VEnc = vEnc,
        };
    }

    /// <summary>Constant-time check of the peer's confirmation MAC.</summary>
    public static bool VerifyPeerConfirm(Spake2PlusOutcome outcome, byte[] received)
        => CryptographicOperations.FixedTimeEquals(outcome.ExpectedPeerConfirm, received);

    // TT = len(Context)||Context || len(idProver)||idProver || len(idVerifier)||idVerifier
    //   || len(M)||M || len(N)||N || len(shareP)||shareP || len(shareV)||shareV
    //   || len(Z)||Z || len(V)||V || len(w0)||w0     (RFC 9383 §3.3)
    private byte[] BuildTranscript(byte[] shareP, byte[] shareV, byte[] zEnc, byte[] vEnc, BigInteger w0)
    {
        using var ms = new MemoryStream();
        AppendLenPrefixed(ms, _context);
        AppendLenPrefixed(ms, _idProver);
        AppendLenPrefixed(ms, _idVerifier);
        AppendLenPrefixed(ms, M.GetEncoded(false));
        AppendLenPrefixed(ms, N.GetEncoded(false));
        AppendLenPrefixed(ms, shareP);
        AppendLenPrefixed(ms, shareV);
        AppendLenPrefixed(ms, zEnc);
        AppendLenPrefixed(ms, vEnc);
        AppendLenPrefixed(ms, FixedScalar(w0)); // 32-byte big-endian
        return ms.ToArray();
    }

    private static void AppendLenPrefixed(Stream s, byte[] value)
    {
        Span<byte> len = stackalloc byte[8];
        System.Buffers.Binary.BinaryPrimitives.WriteUInt64LittleEndian(len, (ulong)value.Length);
        s.Write(len);
        s.Write(value);
    }

    /// <summary>32-byte big-endian encoding of a scalar (P-256 field width).</summary>
    private static byte[] FixedScalar(BigInteger w)
    {
        var raw = w.ToByteArrayUnsigned(); // big-endian, minimal
        if (raw.Length == 32) return raw;
        if (raw.Length > 32) throw new Spake2Exception("scalar too large");
        var padded = new byte[32];
        Array.Copy(raw, 0, padded, 32 - raw.Length, raw.Length);
        return padded;
    }

    private static ECPoint Validated(ECPoint p)
    {
        var n = p.Normalize();
        if (n.IsInfinity || !n.IsValid())
            throw new Spake2Exception("invalid EC point");
        return n;
    }

    private static BigInteger RandomScalar()
    {
        // Sample uniformly in [1, n). Oversample then reduce to avoid modulo bias.
        Span<byte> buf = stackalloc byte[48]; // > order width
        while (true)
        {
            RandomNumberGenerator.Fill(buf);
            var candidate = new BigInteger(1, buf.ToArray()).Mod(Order);
            if (candidate.SignValue > 0)
                return candidate;
        }
    }

    /// <summary>Parse a hex scalar (test/registration helper).</summary>
    public static BigInteger ScalarFromHex(string hex) => new(hex, 16);

    // --- Registration helpers (SPAKE2+ verifier derivation) -----------------

    /// <summary>Reduce a wide (bias-avoiding) byte string to a scalar mod n.</summary>
    public static BigInteger ReduceToScalar(ReadOnlySpan<byte> wide)
        => new BigInteger(1, wide.ToArray()).Mod(Order);

    /// <summary>Reconstruct a stored scalar (e.g. w0) from its 32-byte encoding.</summary>
    public static BigInteger ScalarFromBytes(ReadOnlySpan<byte> bytes)
        => new BigInteger(1, bytes.ToArray()).Mod(Order);

    /// <summary>32-byte big-endian encoding of a scalar (storage helper).</summary>
    public static byte[] EncodeScalar(BigInteger w) => FixedScalar(w);

    /// <summary>The SPAKE2+ registration record L = w1·P, uncompressed-SEC1 encoded.</summary>
    public static byte[] EncodeL(BigInteger w1) => G.Multiply(w1).Normalize().GetEncoded(false);
}

File: zeus-web/src/remote/spake2plus.ts
// SPDX-License-Identifier: GPL-2.0-or-later
//
// Browser SPAKE2+ (RFC 9383, ciphersuite P256-SHA256-HKDF-SHA256-HMAC-SHA256).
// The prover half of the remote-access password handshake (ADR-0008); the radio
// runs the matching C# verifier (Zeus.Server.Hosting/Remote/Spake2Plus.cs). Both
// are pinned to the RFC 9383 Appendix C test vectors so they interoperate.
//
// EC point math: @noble/curves (audited). Hashing/KDF/MAC: @noble/hashes.

import { p256 } from '@noble/curves/nist.js';
import { sha256 } from '@noble/hashes/sha2.js';
import { hkdf } from '@noble/hashes/hkdf.js';
import { hmac } from '@noble/hashes/hmac.js';

const Point = p256.Point;
const ORDER = Point.Fn.ORDER;

// RFC 9382/9383 constant points (compressed SEC1); discrete log unknown.
const M = Point.fromHex('02886e2f97ace46e55ba9dd7242579f2993b64e16ef3dcab95afd497333d8fa12f');
const N = Point.fromHex('03d8bbd6c639c62937b04d997f38c3770719c629d7014d49a24b4f98baa1292b49');

export type Spake2Role = 'prover' | 'verifier';

export interface Spake2Outcome {
  sharedKey: Uint8Array;
  localConfirm: Uint8Array;
  expectedPeerConfirm: Uint8Array;
  // Intermediates, exposed for RFC-vector tests.
  transcript: Uint8Array;
  kMain: Uint8Array;
  kConfirmP: Uint8Array;
  kConfirmV: Uint8Array;
  confirmP: Uint8Array;
  confirmV: Uint8Array;
  zEnc: Uint8Array;
  vEnc: Uint8Array;
}

export class Spake2Plus {
  private scalar?: bigint;
  private share?: Uint8Array;
  private w0?: bigint;
  private w1?: bigint;
  private bigL?: InstanceType<typeof Point>;

  constructor(
    private readonly role: Spake2Role,
    private readonly context: Uint8Array,
    private readonly idProver: Uint8Array,
    private readonly idVerifier: Uint8Array,
  ) {}

  /** Prover (client) share: shareP = x·P + w0·M. `x` is for tests; omit for random. */
  startProver(w0: bigint, w1: bigint, x?: bigint): Uint8Array {
    if (this.role !== 'prover') throw new Error('role mismatch');
    this.w0 = w0;
    this.w1 = w1;
    this.scalar = x ?? randomScalar();
    this.share = Point.BASE.multiply(this.scalar).add(M.multiply(w0)).toBytes(false);
    return this.share;
  }

  /** Verifier (radio) share: shareV = y·P + w0·N. Mainly for the in-TS round-trip test. */
  startVerifier(w0: bigint, encodedL: Uint8Array, y?: bigint): Uint8Array {
    if (this.role !== 'verifier') throw new Error('role mismatch');
    this.w0 = w0;
    this.bigL = decodePoint(encodedL);
    this.scalar = y ?? randomScalar();
    this.share = Point.BASE.multiply(this.scalar).add(N.multiply(w0)).toBytes(false);
    return this.share;
  }

  /** Process the peer share → session key + confirmation MACs. Throws (fails closed) on bad input. */
  process(peerShare: Uint8Array): Spake2Outcome {
    if (this.w0 === undefined || this.scalar === undefined || this.share === undefined)
      throw new Error('start* not called');

    // Capture into locals before any call — TS drops `this.field` narrowing across calls.
    const w0 = this.w0;
    const scalar = this.scalar;
    const myShare = this.share;
    const peer = decodePoint(peerShare);

    let z: InstanceType<typeof Point>;
    let v: InstanceType<typeof Point>;
    let shareP: Uint8Array;
    let shareV: Uint8Array;

    if (this.role === 'prover') {
      const t = peer.subtract(N.multiply(w0)); // Y − w0·N
      z = t.multiply(scalar); // x·(Y − w0·N)
      v = t.multiply(this.w1!); // w1·(Y − w0·N)
      shareP = myShare;
      shareV = peer.toBytes(false);
    } else {
      const t = peer.subtract(M.multiply(w0)); // X − w0·M
      z = t.multiply(scalar); // y·(X − w0·M)
      v = this.bigL!.multiply(scalar); // y·L
      shareP = peer.toBytes(false);
      shareV = myShare;
    }

    const zEnc = z.toBytes(false);
    const vEnc = v.toBytes(false);

    const transcript = buildTranscript(
      this.context, this.idProver, this.idVerifier,
      M.toBytes(false), N.toBytes(false),
      shareP, shareV, zEnc, vEnc, scalarTo32(w0),
    );
    const kMain = sha256(transcript);

    const confirmKeys = hkdf(sha256, kMain, undefined, utf8('ConfirmationKeys'), 64);
    const kConfirmP = confirmKeys.slice(0, 32);
    const kConfirmV = confirmKeys.slice(32);
    const sharedKey = hkdf(sha256, kMain, undefined, utf8('SharedKey'), 32);

    // RFC 9383 §3.4: confirmP = MAC(K_confirmP, shareV); confirmV = MAC(K_confirmV, shareP).
    const confirmP = hmac(sha256, kConfirmP, shareV);
    const confirmV = hmac(sha256, kConfirmV, shareP);

    const [localConfirm, expectedPeerConfirm] =
      this.role === 'prover' ? [confirmP, confirmV] : [confirmV, confirmP];

    return {
      sharedKey, localConfirm, expectedPeerConfirm,
      transcript, kMain, kConfirmP, kConfirmV, confirmP, confirmV, zEnc, vEnc,
    };
  }
}

/** Constant-time confirmation-MAC check. */
export function verifyPeerConfirm(outcome: Spake2Outcome, received: Uint8Array): boolean {
  const a = outcome.expectedPeerConfirm;
  if (a.length !== received.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a[i]! ^ received[i]!;
  return diff === 0;
}

// --- helpers ---------------------------------------------------------------

function decodePoint(bytes: Uint8Array): InstanceType<typeof Point> {
  const p = Point.fromBytes(bytes);
  p.assertValidity(); // reject off-curve / identity — mandatory for PAKE security
  return p;
}

// TT = len(Context)||Context || len(idProver)||idProver || len(idVerifier)||idVerifier
//   || len(M)||M || len(N)||N || len(shareP)||shareP || len(shareV)||shareV
//   || len(Z)||Z || len(V)||V || len(w0)||w0     (RFC 9383 §3.3; 8-byte LE lengths)
function buildTranscript(...fields: Uint8Array[]): Uint8Array {
  const parts: Uint8Array[] = [];
  for (const f of fields) {
    parts.push(u64le(f.length));
    parts.push(f);
  }
  return concat(parts);
}

function u64le(n: number): Uint8Array {
  const out = new Uint8Array(8);
  let v = BigInt(n);
  for (let i = 0; i < 8; i++) {
    out[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return out;
}

function scalarTo32(w: bigint): Uint8Array {
  const out = new Uint8Array(32);
  let v = w;
  for (let i = 31; i >= 0; i--) {
    out[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  if (v !== 0n) throw new Error('scalar too large');
  return out;
}

export function reduceToScalar(wide: Uint8Array): bigint {
  let v = 0n;
  for (const b of wide) v = (v << 8n) | BigInt(b);
  return v % ORDER;
}

function randomScalar(): bigint {
  const buf = new Uint8Array(48); // oversample → reduce, low bias
  while (true) {
    crypto.getRandomValues(buf);
    const s = reduceToScalar(buf);
    if (s > 0n) return s;
  }
}

function utf8(s: string): Uint8Array {
  return new TextEncoder().encode(s);
}

function concat(parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 0);
  const out = new Uint8Array(total);
  let off = 0;
  for (const p of parts) {
    out.set(p, off);
    off += p.length;
  }
  return out;
}
